lib/commons/dom/create-grid.js

```diff
@@ -254,65 +254,61 @@ function isFlexOrGridContainer(vNode) {
-    } else if (vNode.getComputedStylePropertyValue('float') !== 'none') {
-      // Put floated elements above z-index: 0
-      // (step #5 floating get sorted below step #8 positioned)
-      stackingOrder.push(createContext(FLOAT_ORDER, vNode));
-    }
-    return stackingOrder;
-  }
-
   // if an element creates a stacking context, find the first
   // true stack (not a "fake" stack created from positioned or
   // floated elements without a z-index) and create a new stack at
   // that point (step #5 and step #8)
   // @see https://www.w3.org/Style/css2-updates/css2/zindex.html
-  const index = stackingOrder.findIndex(({ value }) =>
-    [ROOT_ORDER, FLOAT_ORDER, POSITION_STATIC_ORDER].includes(value)
-  );
-  if (index !== -1) {
-    stackingOrder.splice(index, stackingOrder.length - index);
+  if (isStackingContext(vNode, parentVNode)) {
+    const index = stackingOrder.findIndex(({ value }) =>
+      [ROOT_ORDER, FLOAT_ORDER, POSITION_ORDER].includes(value)
+    );
+    if (index !== -1) {
+      stackingOrder.splice(index, stackingOrder.length - index);
+    }
   }
 
   const zIndex = getRealZIndex(vNode, parentVNode);
   if (!['auto', '0'].includes(zIndex)) {
-    stackingOrder.push(createContext(parseInt(zIndex), vNode));
+    stackingOrder.push(createContext(parseInt(zIndex), vNode, index));
+    return stackingOrder;
+  }
+
+  // if a positioned element has z-index: auto or 0 (step #8), or if
+  // a non-positioned floating element (step #5), treat it as its
+  // own stacking context
+  // @see https://www.w3.org/Style/css2-updates/css2/zindex.html
+  if (vNode.getComputedStylePropertyValue('position') !== 'static') {
+    // Put positioned elements above floated elements
+    stackingOrder.push(createContext(POSITION_ORDER, vNode, index));
+    return stackingOrder;
+  }
+
+  if (vNode.getComputedStylePropertyValue('float') !== 'none') {
+    // Put floated elements above z-index: 0
+    // (step #5 floating get sorted below step #8 positioned)
+    stackingOrder.push(createContext(FLOAT_ORDER, vNode, index));
+    return stackingOrder;
+  }
+
+  if (!isStackingContext(vNode, parentVNode)) {
     return stackingOrder;
   }
-  // since many things can create a new stacking context without position or
-  // z-index, we need to know the order in the dom to sort them by. Use the
-  // nodeIndex property to create a number less than the "fake" stacks from
-  // positioned or floated elements but still larger than 0
-  // 10 pad gives us the ability to sort up to 1B nodes (padStart does not
-  // exist in ie11)
-  let float = nodeIndex.toString();
-  while (float.length < 10) {
-    float = '0' + float;
-  }
-  stackingOrder.push(
-    createContext(parseFloat(`${DEFAULT_ORDER}${float}`), vNode)
-  );
 
+  stackingOrder.push(createContext(DEFAULT_ORDER, vNode, index));
   return stackingOrder;
 }
 
-function createContext(value, vNode) {
+function createContext(value, vNode, index) {
   return {
     value,
+    nodeIndex: index,
```

Comment:
This tripped me up. Since VirtualNode also has a nodeIndex prop, I think we should call this something different. Are you okay with using `gridIndex` instead?

```suggestion
    gridIndex,
```

Reply:
I agree with avoiding the name conflict with virtualNodes, but I'm not a fan of `gridIndex`; that to me implies that it's an index into the grid, which isn't accurate.

Does the nodeIndex on VirtualNode already have the same semantics? If so, can this just be dropped and have `visuallySort` use that, since it already has references to virtual nodes? If not, maybe `domPosition`, `domOrder`, `domOrderComparator`, or similar?

Reply:
Since the grid can [create elements that are missing from the virtual tree](https://github.com/dequelabs/axe-core/blob/develop/lib/commons/dom/create-grid.js#L80-L82) this `nodeIndex` is not guaranteed to be the same as the one on the virtual tree so I can't use that for sorting. I can call it `domPosition` as it makes it relate to the `compareDocumentPosition` function in my mind.

Reply:
I think I'll call this `treeOrder` as it matches the [spec definition](https://www.w3.org/Style/css2-updates/css2/zindex.html)

> Tree Order
> Preorder depth-first traversal of the rendering tree, in logical (not visual) order for bidirectional content, after taking into account properties that move boxes around.